File: app/src/main/java/com/donygeorge/todolist/MainActivity.java
package com.donygeorge.todolist;

import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.raizlabs.android.dbflow.sql.language.SQLite;

import java.util.ArrayList;
import java.util.List;

import static com.donygeorge.todolist.R.id.itemsListView;

public class MainActivity extends AppCompatActivity implements EditItemDialogFragment.EditItemDialogListener {
    private ArrayList<Item> mItems;
    private ItemsAdapter mItemsAdapter;
    private ListView mItemsListView;
    private final int EDIT_REQUEST_CODE = 20;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        readItems();
        mItemsListView = (ListView)findViewById(itemsListView);
        mItemsAdapter = new ItemsAdapter(this, mItems);
        mItemsListView.setAdapter(mItemsAdapter);
        setupListViewListener();
    }

    public void onAddItem(View v) {
        showEditItemDialogFragment(null, -1);
    }

    private EditItemDialogFragment showEditItemDialogFragment(Item item, int position) {
        FragmentManager fm = getSupportFragmentManager();
        EditItemDialogFragment editItemDialogFragment = EditItemDialogFragment.newInstance(item, position);
        editItemDialogFragment.show(fm, "fragment_edit_item");
        return editItemDialogFragment;
    }

    @Override
    public void onFinishEditDialog(Item item, int position) {
        if (position == -1) {
            mItems.add(item);
        } else {
            mItems.set(position, item);
        }
        mItemsAdapter.notifyDataSetChanged();;
        writeItems();
    }


    private void setupListViewListener() {
        mItemsListView.setOnItemClickListener(
                new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                        showEditItemDialogFragment(mItems.get(i), i);
                    }
                }
        );


        mItemsListView.setOnItemLongClickListener(
                new AdapterView.OnItemLongClickListener() {
                    @Override
                    public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {
                        mItems.remove(i);
                        mItemsAdapter.notifyDataSetChanged();
                        writeItems();
                        return true;
                    }
                }
        );
    }

    private void readItems() {
        List<ItemModel> itemModels = SQLite.select().
                from(ItemModel.class).
                queryList();
        mItems = new ArrayList<Item>();
        for (ItemModel itemModel : itemModels) {
            mItems.add(new Item(itemModel));
        }
    }

    private ItemModel getModel(int position, Item item) {
        ItemModel model = new ItemModel();
        model.id = position;
        model.text = item.text;
        model.date = item.completionDate;
        model.completed = item.completed;
        model.priority = item.priority.ordinal();
        return model;
    }

    private void writeItems() {
        for (int i = 0; i < mItems.size(); i++) {
            ItemModel model = getModel(i, mItems.get(i));
            model.save();
        }

        // Delete extra items
        List<ItemModel> models = SQLite.select().
                from(ItemModel.class).
                queryList();
        for (ItemModel model : models) {
            if (model.id >= mItems.size()) {
                model.delete();
            }
        }
    }
}

File: app/src/main/java/com/donygeorge/todolist/MainActivity.kt
package com.donygeorge.todolist

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.ListView
import com.donygeorge.todolist.R.id.itemsListView
import org.apache.commons.io.FileUtils
import java.io.File
import java.io.IOException
import java.util.*

class MainActivity : AppCompatActivity() {
    private var mItems: ArrayList<String>? = null
    private var mItemsAdapter: ArrayAdapter<String>? = null
    private var mItemsListView: ListView? = null
    private val EDIT_REQUEST_CODE = 20


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        readItems()
        mItemsListView = findViewById(itemsListView) as ListView
        mItemsAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, mItems!!)
        mItemsListView!!.adapter = mItemsAdapter
        setupListViewListener()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        // REQUEST_CODE is defined above
        if (resultCode == Activity.RESULT_OK && requestCode == EDIT_REQUEST_CODE) {
            val text = data.extras.getString("text")
            val position = data.extras.getInt("index", -1)
            mItems!![position] = text
            mItemsAdapter!!.notifyDataSetChanged()
            writeItems()
        }
    }

    fun onAddItem(v: View) {
        val addItemEditText = findViewById(R.id.addItemEditText) as EditText
        val itemText = addItemEditText.text.toString()
        mItemsAdapter!!.add(itemText)
        addItemEditText.setText("")
        writeItems()
    }

    private fun setupListViewListener() {
        mItemsListView!!.onItemClickListener = AdapterView.OnItemClickListener { adapterView, view, i, l ->
            val intent = Intent(this@MainActivity, EditItemActivity::class.java)
            intent.putExtra("index", i)
            intent.putExtra("text", mItems!![i])
            startActivityForResult(intent, EDIT_REQUEST_CODE)
        }


        mItemsListView!!.onItemLongClickListener = AdapterView.OnItemLongClickListener { adapterView, view, i, l ->
            mItems!!.removeAt(i)
            mItemsAdapter!!.notifyDataSetChanged()
            writeItems()
            true
        }
    }

    private fun readItems() {
        val filesDir = filesDir
        val todoFile = File(filesDir, "todo.txt")
        try {
            mItems = ArrayList(FileUtils.readLines(todoFile))
        } catch (e: IOException) {
            mItems = ArrayList<String>()
        }
    }

    private fun writeItems() {
        val filesDir = filesDir
        val todoFile = File(filesDir, "todo.txt")
        try {
            FileUtils.writeLines(todoFile, mItems)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
